### DISCUSSION TITLE
How can I replace or add a Video in Android

### DISCUSSION THREAD

**Discussion by Meet-Miyani** [Tencent/libpag#469] [General] [answered]
![image](https://user-images.githubusercontent.com/91192898/182446199-267bd85d-e75a-467d-9601-8b4085a1599c.png)
Here is the image of **PAG-Player Windows**, and it allows us to select any media for the layer, for example either **Image or Video**. It works fine in the windows, but I didn't found any method to replace or add the video in the Android Library. I don't want to convert the video into BMP for Android. As it works in Desktop Player, can I get the same in mobile device ?

**Is it possible by any means ?**

**Comment by lvpengwei** [comment]
You need to decode each frame of the video and pass it to the PAG SDK through the interface of [replaceImage](https://github.com/Tencent/libpag/blob/v4.0.5.17/android/libpag/src/main/java/org/libpag/PAGFile.java#L89).

**Reply by Meet-Miyani** [reply]
You mean each frame into Bitmap ?
But decoding each frame and passing every bitmap from list, it might CPU intensive task,
Storing all bitmaps of a video in a list. Is there any alternative solution to it ? Or can pass a surface directly, I don't have much knowledge of decoding and encoding, or OpenGL.

**Reply by lvpengwei** [reply]
You can decode the video to a surface, make a texture from the surface, then make a PAGImage.

**Reply by domchen** [reply]
Hi @Meet-Miyani, PAG is just a rendering library for AE animations, which can be easily used in video-editing tasks, but it is NOT a video-editing library. You need a third-party video-editing library to do the work. Besides, the knowledge of decoding, encoding, and OpenGL are indeed required for video-editing tasks.

**Reply by Meet-Miyani** [reply]
Yes, I got it @domchen . Would you recommend me any third-party libraries ?
I'll appreciate that.

**Reply by domchen** [reply, marked as answer]
@Meet-Miyani The new version of PAG (4.2) now includes built-in support for video replacement and audio playback abilities. You can view a sample of these advanced features at the following link: https://github.com/libpag/pag-enterprise-android. However, please note that these features are only available in the enterprise version of PAG, which is a paid version.

**Reply by Meet-Miyani** [reply]
Thanks for the support @domchen, can I know the pricing of this paid version? I checked the website but could not find the pricing section.

**Reply by kevingpqi123** [reply]
You can contact our support team via email on the page below for Enterprise Edition purchases.
https://pag.io/product.html#pag-enterprise-edition
<img width="1509" height="657" alt="image" src="https://github.com/user-attachments/assets/1de89ff5-17b0-4d44-8c6f-b85df7070cf8" />
